### PR TITLE
Added information about providing storage to an etcd node using PCI passthrough

### PR DESCRIPTION
**Pull request by ahardin-rh**
https://trello.com/c/IpPgVp1h/408-1-docs-add-a-blurb-to-etcd-section-on-how-to-pass-through-an-ssd-nvme-device-when-running-on-openstack

Preview build: http://file.rdu.redhat.com/~ahardin/10032017/etcd-openstack/scaling_performance/host_practices.html#providing-storage-to-an-etcd-node-using-pci-passthrough

### CODE REVIEW

File: scaling_performance/host_practices.adoc [hunk 1]
```diff
@@ -136,6 +136,62 @@ endif::[]
 etcd processes are typically memory intensive. Master / API server processes are CPU intensive. This makes them a reasonable co-location pair within a single machine or virtual machine (VM). Optimize communication between etcd and master hosts either by co-locating them on the same host, or providing a dedicated network.
 ====
 
+[[providing-storage-to-an-etcd-node-using-pci-passthrough]]
```
Comment (smalleni) [inline scaling_performance/host_practices.adoc]: Do we want to mention that the procedure that follows is OpenStack specific?

File: scaling_performance/host_practices.adoc [hunk 2]
```diff
@@ -136,6 +136,62 @@ endif::[]
...
+*_/etc/sysconfig/grub_* file, add `intel_iommu=on iommu=pt` to the end of the
+`GRUB_CMDLINX_LINUX` line, within the quotation marks.
+
+. Regenerate *_/etc/grub2.cfg_* by running:
++
+----
+$ grub2-mkconfig -o /etc/grub2.cfg
+----
+
+. Reboot the system.
```
Comment (smalleni) [inline scaling_performance/host_practices.adoc]: Above steps need to be done only on the OpenStack Compute Nodes where the PCI device exists

File: scaling_performance/host_practices.adoc [hunk 3]
```diff
@@ -136,6 +136,62 @@ endif::[]
 etcd processes are typically memory intensive. Master / API server processes are CPU intensive. This makes them a reasonable co-location pair within a single machine or virtual machine (VM). Optimize communication between etcd and master hosts either by co-locating them on the same host, or providing a dedicated network.
 ====
 
+[[providing-storage-to-an-etcd-node-using-pci-passthrough]]
+=== Providing Storage to an etcd Node Using PCI Passthrough
+
+To provide fast storage to an etcd node so that etcd is stable at large scale,
+use PCI passthrough. To set this up with Red Hat OpenStack 11 or later:
```
Comment (smalleni) [inline scaling_performance/host_practices.adoc]: use PCI passthrough to pass an NVMe device directly to the etcd node

**Comment by ahardin-rh** [comment]
@jeremyeder PTAL. Also asking Sai to review.

File: scaling_performance/host_practices.adoc [hunk 4]
```diff
@@ -136,6 +136,62 @@ endif::[]
...
+
+enabled_filters=RetryFilter,AvailabilityZoneFilter,RamFilter,DiskFilter,ComputeFilter,ComputeCapabilitiesFilter,ImagePropertiesFilter,ServerGroupAntiAffinityFilter,ServerGroupAffinityFilter,PciPassthroughFilter
+
+available_filters=nova.scheduler.filters.all_filters
+
+[pci]
+
+alias = { "vendor_id":"144d", "product_id":"a820",
+"device_type":"type-PCI", "name":"nvme" }
+----
```
Comment (smalleni) [inline scaling_performance/host_practices.adoc]: Restart nova-api and nova-scheduler on the controllers after this.

File: scaling_performance/host_practices.adoc [hunk 5]
```diff
@@ -136,6 +136,62 @@ endif::[]
...
+. On compute nodes in *_/etc/nova/nova.conf_*:
++
+----
+[pci]
+
+passthrough_whitelist = { "address": "0000:06:00.0" }
+
+alias = { "vendor_id":"144d", "product_id":"a820",
+"device_type":"type-PCI", "name":"nvme" }
+----
```
Comment (smalleni) [inline scaling_performance/host_practices.adoc]: Restart nova-compute on the compute nodes.

File: scaling_performance/host_practices.adoc [hunk 6]
```diff
@@ -136,6 +136,62 @@ endif::[]
...
++
+To retrieve the required `address`, `vendor_id`, and `product_id` values, run:
```
Comment (smalleni) [inline scaling_performance/host_practices.adoc]: To retrieve the required `address`, `vendor_id`, and `product_id` values of the NVMe device you want to passthrough, run:

File: scaling_performance/host_practices.adoc [hunk 7]
```diff
@@ -136,6 +136,62 @@ endif::[]
...
++
+----
+# lspci -nn | grep devicename
+----
+
+For more information on PCI passthrough, see the
+link:https://docs.openstack.org/nova/pike/admin/pci-passthrough.html[Red Hat
+OpenStack documentation].
```
Comment (smalleni) [inline scaling_performance/host_practices.adoc]: This is upstream OpenStack documentation, not specific to Red Hat.

File: scaling_performance/host_practices.adoc [hunk 8]
```diff
@@ -136,6 +136,62 @@ endif::[]
...
+alias = { "vendor_id":"144d", "product_id":"a820",
+"device_type":"type-PCI", "name":"nvme" }
+----
++
+To retrieve the required `address`, `vendor_id`, and `product_id` values, run:
++
+----
+# lspci -nn | grep devicename
+----
+
```
Comment (smalleni) [inline scaling_performance/host_practices.adoc]: Although the OpenStack docs contain more details as stated later, let us just mention that the user needs to configure and OpenStack flavor to use the NVMe and launch the etcd node using this flavor.

**Review by smalleni — CHANGES_REQUESTED** [review]
Few comments inline...

**Comment by ahardin-rh** [comment]
@smalleni Thanks! Updated.

File: scaling_performance/host_practices.adoc [hunk 9]
```diff
@@ -136,6 +136,68 @@ endif::[]
...
+NVMe device you want to passthrough, run:
++
+----
+# lspci -nn | grep devicename
+----
+
+. Restart `nova-compute` on the compute nodes.
+
+. Configure the OpenStack version you are running to use the NVMe and launch the
+etcd node with the same OpenStack version.
```
Comment (smalleni) [inline scaling_performance/host_practices.adoc]: same OpenStack version? Not sure why we need to emphasize on that?

Comment (ahardin-rh) [reply scaling_performance/host_practices.adoc]: @smalleni Oh, I think I misunderstood your previous comment using the word "flavor". I thought there was emphasis where there wasn't any. Perhaps I can rewrite it to just to say:

Configure the OpenStack version you are running to use the NVMe and launch the etcd node.

Comment (ahardin-rh) [reply scaling_performance/host_practices.adoc]: Thanks! This is updated.

**Review by smalleni — APPROVED** [review]

**Comment by ahardin-rh** [comment]
@openshift/team-documentation Please peer review

**Comment by bmcelvee** [comment]
LGTM 🍂

**Comment by ahardin-rh** [comment]
[rev_history]
|xref:../scaling_performance/host_practices.adoc#scaling-performance-capacity-host-practices[Recommended Host Practices]
|Added a new xref:../scaling_performance/host_practices.adoc#providing-storage-to-an-etcd-node-using-pci-passthrough-with-openstack[Providing Storage to an etcd Node Using PCI Passthrough with OpenStack] section.
%

**Pull request merged**